portfolio/src/main/java/com/google/sps/servlets/SpotifyServlet.java

```diff
@@ -0,0 +1,150 @@
+    .scope("user-read-currently-playing")
+    .show_dialog(true)
+    .build();
+ 
+  private static AuthorizationCodeRequest authorizationCodeRequest = spotifyApi.authorizationCode(oauthToken)
+    .build();
+
+  private static GetUsersCurrentlyPlayingTrackRequest getUsersCurrentlyPlayingTrackRequest = spotifyApi
+    .getUsersCurrentlyPlayingTrack()
+    .build();
+
+  private static AuthorizationCodeRefreshRequest authorizationCodeRefreshRequest = spotifyApi.authorizationCodeRefresh()
+    .build();
+
+  public static String authorizationCodeUri() {
+    final URI uri = authorizationCodeUriRequest.execute();
+    return uri.toString();
+  }
+
+  public static String formatAsMinutesAndSeconds(Integer milliseconds) {
+    Integer seconds = (Integer)(milliseconds/ 1000) % 60;
+    Integer minutes = (Integer)(milliseconds/ (1000) / 60);
+    return String.format("%d:%02d", minutes, seconds);
+  }
+
+  public static void authorizationCode() {
+    try {
+      final AuthorizationCodeCredentials authorizationCodeCredentials = authorizationCodeRequest.execute();
+      spotifyApi.setAccessToken(authorizationCodeCredentials.getAccessToken());
+      spotifyApi.setRefreshToken(authorizationCodeCredentials.getRefreshToken());
+      System.out.println("Expires in: " + authorizationCodeCredentials.getExpiresIn());
+    } catch (IOException | SpotifyWebApiException | ParseException e) {
+      System.out.println("Error: " + e.getMessage());
+    }
+  }
+
+  public static void authorizationCodeRefresh() {
+    try {
+      final AuthorizationCodeCredentials authorizationCodeCredentials = authorizationCodeRefreshRequest.execute();
+ 
+      spotifyApi.setAccessToken(authorizationCodeCredentials.getAccessToken());
+      spotifyApi.setRefreshToken(authorizationCodeCredentials.getRefreshToken());
+ 
+      System.out.println("Expires in: " + authorizationCodeCredentials.getExpiresIn());
+    } catch (IOException | SpotifyWebApiException | ParseException e) {
+      System.out.println("Error: " + e.getMessage());
+    }
+  }
+
+  public static String getCurrentTrack() {
+    try {
+      final CompletableFuture<CurrentlyPlaying> currentlyPlayingFuture = getUsersCurrentlyPlayingTrackRequest.executeAsync();
+
+      final CurrentlyPlaying currentlyPlaying = currentlyPlayingFuture.join();
+
+      if (currentlyPlaying == null) {
+        return "No song currently playing.";
+      }
+
+      IPlaylistItem item = currentlyPlaying.getItem();
+      String songName = "";
+      String artists = "";
+      
+      if (item instanceof Track) {
+        songName = ((Track)item).getName();
+
+        for (int iterator = 0; iterator < ((Track)item).getArtists().length; iterator++) {
+          if (iterator == ((Track)item).getArtists().length - 1) {
+            artists += ((Track)item).getArtists()[iterator].getName();
+            break;
+          }
+          artists += ((Track)item).getArtists()[iterator].getName() + ", ";
+        }
```

Comment:
```suggestion
        String.join(", ", ((Track)item).getArtists());
```

You might have to add an import for the java `String` util: https://docs.oracle.com/javase/8/docs/api/java/lang/String.html

Reply:
That's a lot simpler, thank you for the suggestion!